File: sdk/src/services/analytics.ts
import { PublicKey } from "@solana/web3.js";
import { BaseService } from "./base";
import {
  AgentAccount,
  MessageAccount,
  ChannelAccount,
  EscrowAccount,
  MessageStatus,
  ChannelVisibility,
} from "../types";
import {
  lamportsToSol,
  formatDuration,
  formatBytes,
  getCapabilityNames,
  hasCapability,
} from "../utils";

/**
 * Analytics and insights for agent activities, message patterns, and channel usage
 */

export interface AgentAnalytics {
  totalAgents: number;
  capabilityDistribution: Record<string, number>;
  averageReputation: number;
  topAgentsByReputation: AgentAccount[];
  recentlyActive: AgentAccount[];
}

export interface MessageAnalytics {
  totalMessages: number;
  messagesByStatus: Record<MessageStatus, number>;
  messagesByType: Record<string, number>;
  averageMessageSize: number;
  messagesPerDay: number;
  topSenders: Array<{ agent: PublicKey; messageCount: number }>;
  recentMessages: MessageAccount[];
}

export interface ChannelAnalytics {
  totalChannels: number;
  channelsByVisibility: Record<ChannelVisibility, number>;
  averageParticipants: number;
  mostPopularChannels: ChannelAccount[];
  totalEscrowValue: number;
  averageChannelFee: number;
}

export interface NetworkAnalytics {
  totalTransactions: number;
  totalValueLocked: number;
  activeAgents24h: number;
  messageVolume24h: number;
  networkHealth: "healthy" | "moderate" | "congested";
  peakUsageHours: number[];
}

export interface DashboardData {
  agents: AgentAnalytics;
  messages: MessageAnalytics;
  channels: ChannelAnalytics;
  network: NetworkAnalytics;
  generatedAt: number;
}

export class AnalyticsService extends BaseService {
  /**
   * Get comprehensive analytics dashboard
   */
  async getDashboard(): Promise<DashboardData> {
    const [agents, messages, channels, network] = await Promise.all([
      this.getAgentAnalytics(),
      this.getMessageAnalytics(),
      this.getChannelAnalytics(),
      this.getNetworkAnalytics(),
    ]);

    return {
      agents,
      messages,
      channels,
      network,
      generatedAt: Date.now(),
    };
  }

  /**
   * Get agent ecosystem analytics
   */
  async getAgentAnalytics(limit: number = 100): Promise<AgentAnalytics> {
    try {
      const agents = await this.connection.getProgramAccounts(this.programId, {
        filters: [
          {
            memcmp: {
              offset: 0,
              bytes: this.getDiscriminator("agentAccount"),
            },
          },
        ],
        commitment: this.commitment,
      });

      const agentData: AgentAccount[] = agents.map((acc) => {
        const account = this.ensureInitialized().coder.accounts.decode(
          "agentAccount",
          acc.account.data,
        );
        return {
          pubkey: acc.pubkey,
          capabilities: account.capabilities.toNumber(),
          metadataUri: account.metadataUri,
          reputation: account.reputation?.toNumber() || 0,
          lastUpdated: account.lastUpdated?.toNumber() || Date.now(),
          bump: account.bump,
        };
      });

      // Calculate capability distribution
      const capabilityDistribution: Record<string, number> = {};
      agentData.forEach((agent) => {
        const capabilities = getCapabilityNames(agent.capabilities);
        capabilities.forEach((cap) => {
          capabilityDistribution[cap] = (capabilityDistribution[cap] || 0) + 1;
        });
      });

      // Calculate average reputation
      const averageReputation =
        agentData.length > 0
          ? agentData.reduce((sum, agent) => sum + agent.reputation, 0) /
            agentData.length
          : 0;

      // Get top agents by reputation
      const topAgentsByReputation = agentData
        .sort((a, b) => b.reputation - a.reputation)
        .slice(0, 10);

      // Get recently active agents (last 24 hours)
      const twentyFourHoursAgo = Date.now() - 24 * 60 * 60 * 1000;
      const recentlyActive = agentData
        .filter((agent) => agent.lastUpdated * 1000 > twentyFourHoursAgo)
        .sort((a, b) => b.lastUpdated - a.lastUpdated)
        .slice(0, 20);

      return {
        totalAgents: agentData.length,
        capabilityDistribution,
        averageReputation,
        topAgentsByReputation,
        recentlyActive,
      };
    } catch (error: any) {
      throw new Error(`Failed to get agent analytics: ${error.message}`);
    }
  }

  /**
   * Get message analytics and patterns
   */
  async getMessageAnalytics(limit: number = 1000): Promise<MessageAnalytics> {
    try {
      const messages = await this.connection.getProgramAccounts(
        this.programId,
        {
          filters: [
            {
              memcmp: {
                offset: 0,
                bytes: this.getDiscriminator("messageAccount"),
              },
            },
          ],
          commitment: this.commitment,
        },
      );

      const messageData: MessageAccount[] = messages
        .slice(0, limit)
        .map((acc) => {
          const account = this.ensureInitialized().coder.accounts.decode(
            "messageAccount",
            acc.account.data,
          );
          return {
            pubkey: acc.pubkey,
            sender: account.sender,
            recipient: account.recipient,
            payload: account.payload || "",
            payloadHash: account.payloadHash,
            messageType: this.convertMessageTypeFromProgram(
              account.messageType,
            ),
            status: this.convertMessageStatusFromProgram(account.status),
            timestamp: account.timestamp?.toNumber() || Date.now(),
            createdAt: account.createdAt?.toNumber() || Date.now(),
            expiresAt: account.expiresAt?.toNumber() || 0,
            bump: account.bump,
          };
        });

      // Group messages by status
      const messagesByStatus: Record<MessageStatus, number> = {
        [MessageStatus.Pending]: 0,
        [MessageStatus.Delivered]: 0,
        [MessageStatus.Read]: 0,
        [MessageStatus.Failed]: 0,
      };
      messageData.forEach((msg) => {
        messagesByStatus[msg.status]++;
      });

      // Group messages by type
      const messagesByType: Record<string, number> = {};
      messageData.forEach((msg) => {
        const type = msg.messageType.toString();
        messagesByType[type] = (messagesByType[type] || 0) + 1;
      });

      // Calculate average message size
      const averageMessageSize =
        messageData.length > 0
          ? messageData.reduce((sum, msg) => sum + msg.payload.length, 0) /
            messageData.length
          : 0;

      // Calculate messages per day (last 7 days)
      const sevenDaysAgo = Date.now() - 7 * 24 * 60 * 60 * 1000;
      const recentMessages = messageData.filter(
        (msg) => msg.timestamp * 1000 > sevenDaysAgo,
      );
      const messagesPerDay = recentMessages.length / 7;

      // Get top senders
      const senderCounts: Record<string, number> = {};
      messageData.forEach((msg) => {
        const sender = msg.sender.toBase58();
        senderCounts[sender] = (senderCounts[sender] || 0) + 1;
      });

      const topSenders = Object.entries(senderCounts)
        .sort(([, a], [, b]) => b - a)
        .slice(0, 10)
        .map(([agent, messageCount]) => ({
          agent: new PublicKey(agent),
          messageCount,
        }));

      return {
        totalMessages: messageData.length,
        messagesByStatus,
        messagesByType,
        averageMessageSize,
        messagesPerDay,
        topSenders,
        recentMessages: messageData.slice(0, 20),
      };
    } catch (error: any) {
      throw new Error(`Failed to get message analytics: ${error.message}`);
    }
  }

  /**
   * Get channel usage analytics
   */
  async getChannelAnalytics(limit: number = 100): Promise<ChannelAnalytics> {
    try {
      const channels = await this.connection.getProgramAccounts(
        this.programId,
        {
          filters: [
            {
              memcmp: {
                offset: 0,
                bytes: this.getDiscriminator("channelAccount"),
              },
            },
          ],
          commitment: this.commitment,
        },
      );

      const channelData: ChannelAccount[] = channels
        .slice(0, limit)
        .map((acc) => {
          const account = this.ensureInitialized().coder.accounts.decode(
            "channelAccount",
            acc.account.data,
          );
          return {
            pubkey: acc.pubkey,
            creator: account.creator,
            name: account.name,
            description: account.description,
            visibility: this.convertChannelVisibilityFromProgram(
              account.visibility,
            ),
            maxParticipants: account.maxParticipants,
            participantCount: account.currentParticipants,
            currentParticipants: account.currentParticipants,
            feePerMessage: account.feePerMessage?.toNumber() || 0,
            escrowBalance: account.escrowBalance?.toNumber() || 0,
            createdAt: account.createdAt?.toNumber() || Date.now(),
            isActive: true,
            bump: account.bump,
          };
        });

      // Group channels by visibility
      const channelsByVisibility: Record<ChannelVisibility, number> = {
        [ChannelVisibility.Public]: 0,
        [ChannelVisibility.Private]: 0,
      };
      channelData.forEach((channel) => {
        channelsByVisibility[channel.visibility]++;
      });

      // Calculate average participants
      const averageParticipants =
        channelData.length > 0
          ? channelData.reduce(
              (sum, channel) => sum + channel.participantCount,
              0,
            ) / channelData.length
          : 0;

      // Get most popular channels by participant count
      const mostPopularChannels = channelData
        .sort((a, b) => b.participantCount - a.participantCount)
        .slice(0, 10);

      // Calculate total escrow value
      const totalEscrowValue = channelData.reduce(
        (sum, channel) => sum + channel.escrowBalance,
        0,
      );

      // Calculate average channel fee
      const averageChannelFee =
        channelData.length > 0
          ? channelData.reduce(
              (sum, channel) => sum + channel.feePerMessage,
              0,
            ) / channelData.length
          : 0;

      return {
        totalChannels: channelData.length,
        channelsByVisibility,
        averageParticipants,
        mostPopularChannels,
        totalEscrowValue,
        averageChannelFee,
      };
    } catch (error: any) {
      throw new Error(`Failed to get channel analytics: ${error.message}`);
    }
  }

  /**
   * Get network-wide analytics
   */
  async getNetworkAnalytics(): Promise<NetworkAnalytics> {
    try {
      // Get recent block performance for network health
      const recentSlots = await this.connection.getRecentPerformanceSamples(10);
      const averageTps =
        recentSlots.length > 0
          ? recentSlots.reduce((sum, slot) => sum + slot.numTransactions, 0) /
            recentSlots.reduce((sum, slot) => sum + slot.samplePeriodSecs, 0)
          : 0;

      // Determine network health based on TPS
      let networkHealth: "healthy" | "moderate" | "congested" = "healthy";
      if (averageTps < 1000) {
        networkHealth = "congested";
      } else if (averageTps < 2000) {
        networkHealth = "moderate";
      }

      // Get total value locked (from escrow accounts)
      const escrows = await this.connection.getProgramAccounts(this.programId, {
        filters: [
          {
            memcmp: {
              offset: 0,
              bytes: this.getDiscriminator("escrowAccount"),
            },
          },
        ],
        commitment: this.commitment,
      });

      const totalValueLocked = escrows.reduce((sum, acc) => {
        try {
          const account = this.ensureInitialized().coder.accounts.decode(
            "escrowAccount",
            acc.account.data,
          );
          return sum + (account.balance?.toNumber() || 0);
        } catch {
          return sum;
        }
      }, 0);

      // Mock data for metrics that require historical tracking
      // In a real implementation, these would be stored in a time-series database
      const peakUsageHours = [9, 10, 11, 14, 15, 16, 20, 21]; // 9-11am, 2-4pm, 8-9pm UTC

      return {
        totalTransactions: recentSlots.reduce(
          (sum, slot) => sum + slot.numTransactions,
          0,
        ),
        totalValueLocked,
        activeAgents24h: 0, // Would need historical tracking
        messageVolume24h: 0, // Would need historical tracking
        networkHealth,
        peakUsageHours,
      };
    } catch (error: any) {
      throw new Error(`Failed to get network analytics: ${error.message}`);
    }
  }

  /**
   * Generate analytics report
   */
  async generateReport(): Promise<string> {
    const dashboard = await this.getDashboard();

    let report = "# PoD Protocol Analytics Report\n\n";
    report += `Generated: ${new Date(dashboard.generatedAt).toISOString()}\n\n`;

    // Agent Analytics
    report += "## Agent Analytics\n";
    report += `- Total Agents: ${dashboard.agents.totalAgents}\n`;
    report += `- Average Reputation: ${dashboard.agents.averageReputation.toFixed(2)}\n`;
    report += `- Recently Active (24h): ${dashboard.agents.recentlyActive.length}\n`;
    report += "\n### Capability Distribution\n";
    Object.entries(dashboard.agents.capabilityDistribution).forEach(
      ([cap, count]) => {
        report += `- ${cap}: ${count} agents\n`;
      },
    );

    // Message Analytics
    report += "\n## Message Analytics\n";
    report += `- Total Messages: ${dashboard.messages.totalMessages}\n`;
    report += `- Average Message Size: ${formatBytes(dashboard.messages.averageMessageSize)}\n`;
    report += `- Messages per Day: ${dashboard.messages.messagesPerDay.toFixed(1)}\n`;
    report += "\n### Message Status Distribution\n";
    Object.entries(dashboard.messages.messagesByStatus).forEach(
      ([status, count]) => {
        report += `- ${status}: ${count} messages\n`;
      },
    );

    // Channel Analytics
    report += "\n## Channel Analytics\n";
    report += `- Total Channels: ${dashboard.channels.totalChannels}\n`;
    report += `- Average Participants: ${dashboard.channels.averageParticipants.toFixed(1)}\n`;
    report += `- Total Value Locked: ${lamportsToSol(dashboard.channels.totalEscrowValue).toFixed(4)} SOL\n`;
    report += `- Average Channel Fee: ${lamportsToSol(dashboard.channels.averageChannelFee).toFixed(6)} SOL\n`;

    // Network Analytics
    report += "\n## Network Analytics\n";
    report += `- Network Health: ${dashboard.network.networkHealth.toUpperCase()}\n`;
    report += `- Total Value Locked: ${lamportsToSol(dashboard.network.totalValueLocked).toFixed(4)} SOL\n`;
    report += `- Peak Usage Hours (UTC): ${dashboard.network.peakUsageHours.join(", ")}\n`;

    return report;
  }

  // ============================================================================
  // Helper Methods
  // ============================================================================

  private getDiscriminator(accountType: string): string {
    // This would need to be implemented based on your IDL
    const discriminators: Record<string, string> = {
      agentAccount: "6RdcqmKGhkRy",
      messageAccount: "6RdcqmKGhkRz",
      channelAccount: "6RdcqmKGhkRA",
      escrowAccount: "6RdcqmKGhkRB",
    };
    return discriminators[accountType] || "";
  }

  private convertMessageTypeFromProgram(programType: any): any {
    if (programType.text !== undefined) return "Text";
    if (programType.data !== undefined) return "Data";
    if (programType.command !== undefined) return "Command";
    if (programType.response !== undefined) return "Response";
    if (programType.custom !== undefined)
      return `Custom(${programType.custom})`;
    return "Unknown";
  }

  private convertMessageStatusFromProgram(programStatus: any): MessageStatus {
    if (programStatus.pending) return MessageStatus.Pending;
    if (programStatus.delivered) return MessageStatus.Delivered;
    if (programStatus.read) return MessageStatus.Read;
    if (programStatus.failed) return MessageStatus.Failed;
    return MessageStatus.Pending;
  }

  private convertChannelVisibilityFromProgram(
    programVisibility: any,
  ): ChannelVisibility {
    if (programVisibility.public !== undefined) return ChannelVisibility.Public;
    if (programVisibility.private !== undefined)
      return ChannelVisibility.Private;
    return ChannelVisibility.Public;
  }
}

File: sdk-python/pod_protocol/services/analytics.py
"""
Analytics service for generating insights on agent activities, message patterns, and channel usage.
"""

import time
import json
from typing import List, Optional, Dict, Any, NamedTuple
from dataclasses import dataclass
from enum import Enum

from .base import BaseService
from ..types import (
    PublicKey, MessageStatus, ChannelVisibility, MessageType,
    AgentAccount, MessageAccount, ChannelAccount, EscrowAccount
)
from ..utils import (
    lamports_to_sol, format_duration, format_bytes, 
    get_capability_names, has_capability
)


@dataclass
class AgentAnalytics:
    """Agent ecosystem analytics."""
    total_agents: int
    capability_distribution: Dict[str, int]
    average_reputation: float
    top_agents_by_reputation: List[AgentAccount]
    recently_active: List[AgentAccount]


@dataclass
class MessageAnalytics:
    """Message analytics and patterns."""
    total_messages: int
    messages_by_status: Dict[str, int]
    messages_by_type: Dict[str, int]
    average_message_size: float
    messages_per_day: float
    top_senders: List[Dict[str, Any]]
    recent_messages: List[MessageAccount]


@dataclass
class ChannelAnalytics:
    """Channel usage analytics."""
    total_channels: int
    channels_by_visibility: Dict[str, int]
    average_participants: float
    most_popular_channels: List[ChannelAccount]
    total_escrow_value: int
    average_channel_fee: float


@dataclass
class NetworkAnalytics:
    """Network-wide analytics."""
    total_transactions: int
    total_value_locked: int
    active_agents_24h: int
    message_volume_24h: int
    network_health: str
    peak_usage_hours: List[int]


@dataclass
class DashboardData:
    """Comprehensive analytics dashboard data."""
    agents: AgentAnalytics
    messages: MessageAnalytics
    channels: ChannelAnalytics
    network: NetworkAnalytics
    generated_at: int


class AnalyticsService(BaseService):
    """Service for analytics and insights."""

    async def get_dashboard(self) -> DashboardData:
        """
        Get comprehensive analytics dashboard.
        
        Returns:
            Complete dashboard data
        """
        agents_analytics = await self.get_agent_analytics()
        messages_analytics = await self.get_message_analytics()
        channels_analytics = await self.get_channel_analytics()
        network_analytics = await self.get_network_analytics()
        
        return DashboardData(
            agents=agents_analytics,
            messages=messages_analytics,
            channels=channels_analytics,
            network=network_analytics,
            generated_at=int(time.time() * 1000)
        )

    async def get_agent_analytics(self, limit: int = 100) -> AgentAnalytics:
        """
        Get agent ecosystem analytics.
        
        Args:
            limit: Maximum number of agents to analyze
            
        Returns:
            Agent analytics data
        """
        try:
            # Get agent accounts
            agents = await self._fetch_agent_accounts(limit)
            
            # Calculate capability distribution
            capability_distribution = {}
            for agent in agents:
                capabilities = get_capability_names(agent.capabilities)
                for cap in capabilities:
                    capability_distribution[cap] = capability_distribution.get(cap, 0) + 1
            
            # Calculate average reputation
            average_reputation = (
                sum(agent.reputation for agent in agents) / len(agents)
                if agents else 0
            )
            
            # Get top agents by reputation
            top_agents_by_reputation = sorted(
                agents, 
                key=lambda x: x.reputation, 
                reverse=True
            )[:10]
            
            # Get recently active agents (last 24 hours)
            twenty_four_hours_ago = int(time.time() * 1000) - (24 * 60 * 60 * 1000)
            recently_active = [
                agent for agent in agents 
                if agent.last_updated * 1000 > twenty_four_hours_ago
            ]
            recently_active.sort(key=lambda x: x.last_updated, reverse=True)
            recently_active = recently_active[:20]
            
            return AgentAnalytics(
                total_agents=len(agents),
                capability_distribution=capability_distribution,
                average_reputation=average_reputation,
                top_agents_by_reputation=top_agents_by_reputation,
                recently_active=recently_active
            )
        except Exception as e:
            raise Exception(f"Failed to get agent analytics: {e}")

    async def get_message_analytics(self, limit: int = 1000) -> MessageAnalytics:
        """
        Get message analytics and patterns.
        
        Args:
            limit: Maximum number of messages to analyze
            
        Returns:
            Message analytics data
        """
        try:
            # Get message accounts
            messages = await self._fetch_message_accounts(limit)
            
            # Group messages by status
            messages_by_status = {}
            for status in MessageStatus:
                messages_by_status[status.value] = sum(
                    1 for msg in messages if msg.status == status
                )
            
            # Group messages by type
            messages_by_type = {}
            for msg in messages:
                msg_type = msg.message_type.value if hasattr(msg.message_type, 'value') else str(msg.message_type)
                messages_by_type[msg_type] = messages_by_type.get(msg_type, 0) + 1
            
            # Calculate average message size
            average_message_size = (
                sum(len(msg.payload or "") for msg in messages) / len(messages)
                if messages else 0
            )
            
            # Calculate messages per day (last 7 days)
            seven_days_ago = int(time.time() * 1000) - (7 * 24 * 60 * 60 * 1000)
            recent_messages = [
                msg for msg in messages 
                if msg.timestamp * 1000 > seven_days_ago
            ]
            messages_per_day = len(recent_messages) / 7.0
            
            # Get top senders
            sender_counts = {}
            for msg in messages:
                sender_str = str(msg.sender)
                sender_counts[sender_str] = sender_counts.get(sender_str, 0) + 1
            
            top_senders = [
                {"agent": PublicKey(sender), "message_count": count}
                for sender, count in sorted(
                    sender_counts.items(), 
                    key=lambda x: x[1], 
                    reverse=True
                )[:10]
            ]
            
            return MessageAnalytics(
                total_messages=len(messages),
                messages_by_status=messages_by_status,
                messages_by_type=messages_by_type,
                average_message_size=average_message_size,
                messages_per_day=messages_per_day,
                top_senders=top_senders,
                recent_messages=messages[:20]
            )
        except Exception as e:
            raise Exception(f"Failed to get message analytics: {e}")

    async def get_channel_analytics(self, limit: int = 100) -> ChannelAnalytics:
        """
        Get channel usage analytics.
        
        Args:
            limit: Maximum number of channels to analyze
            
        Returns:
            Channel analytics data
        """
        try:
            # Get channel accounts
            channels = await self._fetch_channel_accounts(limit)
            
            # Group channels by visibility
            channels_by_visibility = {}
            for visibility in ChannelVisibility:
                channels_by_visibility[visibility.value] = sum(
                    1 for channel in channels if channel.visibility == visibility
                )
            
            # Calculate average participants
            average_participants = (
                sum(channel.participant_count for channel in channels) / len(channels)
                if channels else 0
            )
            
            # Get most popular channels by participant count
            most_popular_channels = sorted(
                channels, 
                key=lambda x: x.participant_count, 
                reverse=True
            )[:10]
            
            # Calculate total escrow value
            total_escrow_value = sum(channel.escrow_balance for channel in channels)
            
            # Calculate average channel fee
            average_channel_fee = (
                sum(channel.fee_per_message for channel in channels) / len(channels)
                if channels else 0
            )
            
            return ChannelAnalytics(
                total_channels=len(channels),
                channels_by_visibility=channels_by_visibility,
                average_participants=average_participants,
                most_popular_channels=most_popular_channels,
                total_escrow_value=total_escrow_value,
                average_channel_fee=average_channel_fee
            )
        except Exception as e:
            raise Exception(f"Failed to get channel analytics: {e}")

    async def get_network_analytics(self) -> NetworkAnalytics:
        """
        Get network-wide analytics.
        
        Returns:
            Network analytics data
        """
        try:
            # Get recent block performance for network health
            # This would typically query the RPC for performance data
            # For now, we'll use mock data
            average_tps = 1500  # Mock TPS
            
            # Determine network health based on TPS
            if average_tps < 1000:
                network_health = "congested"
            elif average_tps < 2000:
                network_health = "moderate"
            else:
                network_health = "healthy"
            
            # Get total value locked (from escrow accounts)
            escrow_accounts = await self._fetch_escrow_accounts()
            total_value_locked = sum(escrow.balance for escrow in escrow_accounts)
            
            # Historical metrics: query for last 24h
            since = int(time.time() * 1000) - (24 * 60 * 60 * 1000)
            
            # Get messages from last 24h for activity metrics
            recent_messages = await self._fetch_recent_messages(since)
            message_volume_24h = len(recent_messages)
            
            # Get unique senders for active agents
            active_agents_24h = len(set(str(msg.sender) for msg in recent_messages))
            
            # Compute peak usage hours
            hour_counts = [0] * 24
            for msg in recent_messages:
                from datetime import datetime
                hour = datetime.fromtimestamp(msg.created_at / 1000).hour
                hour_counts[hour] += 1
            
            peak_usage_hours = [
                hour for hour, count in enumerate(hour_counts) if count > 0
            ]
            
            return NetworkAnalytics(
                total_transactions=len(recent_messages),
                total_value_locked=total_value_locked,
                active_agents_24h=active_agents_24h,
                message_volume_24h=message_volume_24h,
                network_health=network_health,
                peak_usage_hours=peak_usage_hours
            )
        except Exception as e:
            raise Exception(f"Failed to get network analytics: {e}")

    async def generate_report(self) -> str:
        """
        Generate analytics report.
        
        Returns:
            Formatted analytics report
        """
        dashboard = await self.get_dashboard()
        
        report = "# PoD Protocol Analytics Report\n\n"
        report += f"Generated: {time.strftime('%Y-%m-%d %H:%M:%S UTC', time.gmtime(dashboard.generated_at / 1000))}\n\n"
        
        # Agent Analytics
        report += "## Agent Analytics\n"
        report += f"- Total Agents: {dashboard.agents.total_agents}\n"
        report += f"- Average Reputation: {dashboard.agents.average_reputation:.2f}\n"
        report += f"- Recently Active (24h): {len(dashboard.agents.recently_active)}\n"
        report += "\n### Capability Distribution\n"
        for cap, count in dashboard.agents.capability_distribution.items():
            report += f"- {cap}: {count} agents\n"
        
        # Message Analytics
        report += "\n## Message Analytics\n"
        report += f"- Total Messages: {dashboard.messages.total_messages}\n"
        report += f"- Average Message Size: {format_bytes(dashboard.messages.average_message_size)}\n"
        report += f"- Messages per Day: {dashboard.messages.messages_per_day:.1f}\n"
        report += "\n### Message Status Distribution\n"
        for status, count in dashboard.messages.messages_by_status.items():
            report += f"- {status}: {count} messages\n"
        
        # Channel Analytics
        report += "\n## Channel Analytics\n"
        report += f"- Total Channels: {dashboard.channels.total_channels}\n"
        report += f"- Average Participants: {dashboard.channels.average_participants:.1f}\n"
        report += f"- Total Value Locked: {lamports_to_sol(dashboard.channels.total_escrow_value):.4f} SOL\n"
        report += f"- Average Channel Fee: {lamports_to_sol(dashboard.channels.average_channel_fee):.6f} SOL\n"
        
        # Network Analytics
        report += "\n## Network Analytics\n"
        report += f"- Network Health: {dashboard.network.network_health.upper()}\n"
        report += f"- Total Value Locked: {lamports_to_sol(dashboard.network.total_value_locked):.4f} SOL\n"
        report += f"- Peak Usage Hours (UTC): {', '.join(map(str, dashboard.network.peak_usage_hours))}\n"
        
        return report

    # Helper Methods
    async def _fetch_agent_accounts(self, limit: int = 100) -> List[AgentAccount]:
        """Fetch agent accounts from the program."""
        try:
            program = self.ensure_initialized()
            accounts = await program.account.agent_account.all()
            
            result = []
            for acc in accounts[:limit]:
                agent = AgentAccount(
                    pubkey=acc.public_key,
                    capabilities=acc.account.capabilities,
                    metadata_uri=acc.account.metadata_uri,
                    reputation=acc.account.reputation or 0,
                    last_updated=acc.account.last_updated or int(time.time()),
                    invites_sent=acc.account.invites_sent or 0,
                    last_invite_at=acc.account.last_invite_at or 0,
                    bump=acc.account.bump
                )
                result.append(agent)
            
            return result
        except Exception:
            return []

    async def _fetch_message_accounts(self, limit: int = 1000) -> List[MessageAccount]:
        """Fetch message accounts from the program."""
        try:
            program = self.ensure_initialized()
            accounts = await program.account.message_account.all()
            
            result = []
            for acc in accounts[:limit]:
                message = MessageAccount(
                    pubkey=acc.public_key,
                    sender=acc.account.sender,
                    recipient=acc.account.recipient,
                    payload=acc.account.payload or "",
                    payload_hash=acc.account.payload_hash,
                    message_type=self._convert_message_type_from_program(acc.account.message_type),
                    status=self._convert_message_status_from_program(acc.account.status),
                    timestamp=acc.account.timestamp or int(time.time()),
                    created_at=acc.account.created_at or int(time.time()),
                    expires_at=acc.account.expires_at or 0,
                    bump=acc.account.bump
                )
                result.append(message)
            
            return result
        except Exception:
            return []

    async def _fetch_channel_accounts(self, limit: int = 100) -> List[ChannelAccount]:
        """Fetch channel accounts from the program."""
        try:
            program = self.ensure_initialized()
            accounts = await program.account.channel_account.all()
            
            result = []
            for acc in accounts[:limit]:
                channel = ChannelAccount(
                    pubkey=acc.public_key,
                    creator=acc.account.creator,
                    name=acc.account.name,
                    description=acc.account.description,
                    visibility=self._convert_channel_visibility_from_program(acc.account.visibility),
                    max_participants=acc.account.max_participants,
                    participant_count=acc.account.current_participants,
                    current_participants=acc.account.current_participants,
                    fee_per_message=acc.account.fee_per_message or 0,
                    escrow_balance=acc.account.escrow_balance or 0,
                    created_at=acc.account.created_at or int(time.time()),
                    is_active=True,
                    bump=acc.account.bump
                )
                result.append(channel)
            
            return result
        except Exception:
            return []

    async def _fetch_escrow_accounts(self) -> List[EscrowAccount]:
        """Fetch escrow accounts from the program."""
        try:
            program = self.ensure_initialized()
            accounts = await program.account.escrow_account.all()
            
            result = []
            for acc in accounts:
                escrow = EscrowAccount(
                    channel=acc.account.channel,
                    depositor=acc.account.depositor,
                    balance=acc.account.balance or 0,
                    amount=acc.account.balance or 0,
                    created_at=acc.account.created_at or int(time.time()),
                    last_updated=acc.account.last_updated or int(time.time()),
                    bump=acc.account.bump
                )
                result.append(escrow)
            
            return result
        except Exception:
            return []

    async def _fetch_recent_messages(self, since: int) -> List[MessageAccount]:
        """Fetch recent messages since timestamp."""
        try:
            all_messages = await self._fetch_message_accounts()
            return [
                msg for msg in all_messages 
                if msg.created_at * 1000 > since
            ]
        except Exception:
            return []

    def _convert_message_type_from_program(self, program_type: Any) -> MessageType:
        """Convert message type from program format."""
        if isinstance(program_type, dict):
            if "text" in program_type:
                return MessageType.TEXT
            elif "data" in program_type:
                return MessageType.DATA
            elif "command" in program_type:
                return MessageType.COMMAND
            elif "response" in program_type:
                return MessageType.RESPONSE
        return MessageType.TEXT

    def _convert_message_status_from_program(self, program_status: Any) -> MessageStatus:
        """Convert message status from program format."""
        if isinstance(program_status, dict):
            if "pending" in program_status:
                return MessageStatus.PENDING
            elif "delivered" in program_status:
                return MessageStatus.DELIVERED
            elif "read" in program_status:
                return MessageStatus.READ
            elif "failed" in program_status:
                return MessageStatus.FAILED
        return MessageStatus.PENDING

    def _convert_channel_visibility_from_program(self, program_visibility: Any) -> ChannelVisibility:
        """Convert channel visibility from program format."""
        if isinstance(program_visibility, dict):
            if "public" in program_visibility:
                return ChannelVisibility.PUBLIC
            elif "private" in program_visibility:
                return ChannelVisibility.PRIVATE
        return ChannelVisibility.PUBLIC
